doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,369 @@
+Explainable::new(&mut plan).explain(format, context)
+```
+
+Basically
+
+- The `config` is something only depends on the current catalog state.
+- The `humanizer` value can be the current catalog state because item IDs are stable and cannot be recycled.
+- The `finishing` and `duration` values can be constants.
+
+so the only time-dependent information that we need to record is the explained `plan`, which is a `DataflowDescription` instance.
+The next section expands how we plan to do that with more concrete implementation details.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+There are two competing approaches that both follow the same high-level structure.
+
+The [alternatives](#considered-alternatives) presents a design that overall seems to be the obvious solution.
+However, due to some inerent issues that arise when trying to implement it, my proposal is slightly different.
+It will allow us to ship the feature and buy us some time to figure out how to properly refactor the code in order to ultimately move to the (conceptually better) alternative solution.
+Most likely, steps in this direction can be made with MaterializeInc/materialize#20569, but I don't think we should take this as a hard prerequisite for shipping MaterializeInc/materialize#20652.
+
+## Provision plan memoization infrastructure in the `Catalog`
+
+Ideally, we want to provision the structures to hold the `OPTIMIZED` and `PHYSICAL` plans directly in the `CatalogState`.
+However, this turns out to be a non trivial task (see the discussion for the [alternative solution](#considered-alternatives)).
+
+Instead, we propose to directly extend `Catalog` with a new field
+
+```rust
+pub struct Catalog {
+    ...
+    plans: CatalogPlans,
+}
+```
+
+where `CatalogPlans` is an in-memory side container for the derived plans
+
+```rust
+pub struct CatalogPlans {
+    optimized_plan_by_id: BTreeMap<GlobalId, DataflowDesc>,
+    physical_plan_by_id: BTreeMap<GlobalId, DataflowDesc>,
+}
+```
+
+and the `Catalog` exposes the following interface for memoizing and retrieving plans:
+
+```rust
+impl Catalog {
+    pub fn set_optimized_plan(&mut self, id: GlobalId, plan: DataflowDesc) {}
+
+    pub fn set_physical_plan(&mut self, id: GlobalId, plan: DataflowDesc) {}
+
+    pub fn try_get_optimized_plan(&self, id: &GlobalId) -> Option<&DataflowDesc> {}
+
+    pub fn try_get_physical_plan(&self, id: &GlobalId) -> Option<&DataflowDesc> {}
+
+    pub fn drop_plans(&mut self, id: GlobalId)  {}
+}
+```
+
+## Memoize plans
+
+Producing the `OPTIMIZED` and `PHYSICAL`` plans before the `CatalogItem` is constructed comes at considerable refactoring effort (see the [alternative solution](#considered-alternatives) discussion).
+With this solution, we are not forced to commit to this effort now.
+Instead, we can just use the new `Catalog` API and call `set_optimized_plan` and `set_physical_plan` at the stages where we currently produce those plans:
```

Comment:
Add a third entry to the list below discussing dropping, which is also part of the lifecycle. Probably a fourth entry for any ALTER (like the ALTER SET CLUSTER) work.